### PR TITLE
Only execute MT if translation is outdated

### PR DESCRIPTION
**Pull request by JoeyStk**
### Short description

This PR restricts machine translations to translations that are not up-to-date

### Proposed changes

- Check if translation status of target translation is neither UP_TO_DATE nor MACHINE_TRANSLATED
- Add release note

### Side effects

- I think none

### Resolved issues

Fixes: #2247

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

File: integreat_cms/deepl_api/deepl_api_client.py
Comment (MizukiTemma) [inline integreat_cms/deepl_api/deepl_api_client.py:145]:
```suggestion
                if existing_target_translation and existing_target_translation.translation_state in (
```

To avoid `Nonetype Object` error, check existence of translation first then status.

File: integreat_cms/deepl_api/deepl_api_client.py
Comment (MizukiTemma) [inline integreat_cms/deepl_api/deepl_api_client.py:151]:
```suggestion
                        _("There already is an up-to-date translation for {}").format(
                            content_object.best_translation.title
                        ),
```

It's more user friendly to indicate for which content a translation already actual and for which not especially when multiple of contents are translated at once.

**Review by MizukiTemma — CHANGES_REQUESTED** [review]
Thank you for the PR 🙂 This is a nice feature letting users save translation budget.

I tested it with bulk action and have two suggestions.

File: integreat_cms/deepl_api/deepl_api_client.py
Comment (JoeyStk) [reply integreat_cms/deepl_api/deepl_api_client.py]: Oh, that's a very good point. Thank you!

File: integreat_cms/deepl_api/deepl_api_client.py
Comment (JoeyStk) [reply integreat_cms/deepl_api/deepl_api_client.py]: I had wanted to do that too, but couldn't get it to work so pylint wouldn't complain. But this seems to work. Thank you!

**Comment by MizukiTemma** [comment]
MT by checking a box in "implications on other translation" is mentioned in the "Steps to reproduce" in issue description, though in this procedure I think triggering of MT is anyway wished, because it is meant that the content is somehow updated in the source language and the translation of target language to be updated too. But for bulk action the new feature is very helpfull :)

Can we probably introduce this feature also for SUMM.AI, or for SUMM.AI saving budget is not important (due to cost, ect)?

**Comment by timoludwig** [comment]
> Can we probably introduce this feature also for SUMM.AI, or for SUMM.AI saving budget is not important (due to cost, ect)?

I agree, since the added check has nothing to do with DeepL in paarticular, I think the DeepL API client is the wrong place for it, we should rather move this check into the generic MT translation base classes so all MT providers benefit from it.

**Comment by JoeyStk** [comment]
I won't be able to work on it this week anymore, but will move the check to the correct file next week

**Review by seluianova — CHANGES_REQUESTED** [review]
I tested a bit and found one issue. 
Steps to reproduce:

1. Select 2 translations - one up-to-date and one outdated.
2. Execute MT in bulk actions

Result:
Error "There already is an up-to-date translation for {}" is shown for both pages, however only one of them is up-to-date.

I guess the expected result would be:
1 successful translation + 1 error message

File: integreat_cms/core/utils/machine_translation_provider.py
Comment (MizukiTemma) [inline integreat_cms/core/utils/machine_translation_provider.py:219]:
```suggestion
    def is_needed(self, queryset, target_language):
        """
        Checks if a machine translation is needed, thus checking if the
        translation status is UP_TO_DATE or MACHINE_TRANSLATED and then
        returns a lit of translations which are to be updated

        :param queryset: The content model which should be translated
        :type queryset: ~django.db.models.query.QuerySet [ ~integreat_cms.cms.models.abstract_content_model.AbstractContentModel ]

        :param target_language: The target language
        :type target_language: ~integreat_cms.cms.models.languages.language.Language

        :return: translations which need to be translated and updated
        :rtype: list
        """
        # Before translating, check if translation is not up-to-date
        to_translate = []
        for content_object in queryset:
            existing_target_translation = content_object.get_translation(
                target_language.slug
            )
            if (
                existing_target_translation
                and existing_target_translation.translation_state
                in (
                    "UP_TO_DATE",
                    "MACHINE_TRANSLATED",
                )
            ):
                logger.debug(
                    "There already is an up-to-date translation for %s",
                    content_object.best_translation.title,
                )
            else:
                to_translate = to_translate + [content_object]
        return to_translate
```

**Review by MizukiTemma — CHANGES_REQUESTED** [review]
These are suggestions to fix the problem reported by @seluianova 

I couldn't add this suggestion in `bulk_action_views.py` since it was too large.
The following is from line 145 to the end of the post function.

if language_node.mt_provider.bulk_only_for_staff and not request.user.is_staff:
            raise PermissionDenied(
                f"Only staff users have the permission to bulk translate {self.form._meta.model._meta.model_name} via {language_node.mt_provider}"
            )

        to_translate = language_node.mt_provider.is_needed(
            request.region, self.get_queryset(), language_node
        )
        if not to_translate:
            messages.error(
                request,
                _("All the selected translations are already up-to-date."
                ),
            )
            return super().post(request, *args, **kwargs)

        for content_object in self.get_queryset():
            if not content_object in to_translate:
                messages.error(
                    request,
                    _("There already is an up-to-date translation for {}").format(
                        content_object.best_translation.title,
                    ),
                )
        
        logger.debug(
            "Machine translation via %s into %r for: %r",
            language_node.mt_provider.name,
            language_node.language,
            to_translate,
        )
        api_client = language_node.mt_provider.api_client(request, self.form)
        api_client.translate_queryset(to_translate, language_node.slug)

        # Let the base view handle the redirect
        return super().post(request, *args, **kwargs)

**Comment by JoeyStk** [comment]
@seluianova @MizukiTemma Thank you so much for you contributions. 

- The requested changes should be applied now. 
- Following @seluianova's recommendation, I disabled the failing tests for now. As far as I understood the situation this PR is blocked until [Issue #2119](https://github.com/digitalfabrik/integreat-cms/issues/2119) is fixed

**Comment by seluianova** [comment]
> I highly doubt that a race condition is really the cause when the tests are failing consistently

As far as I could see, these failures occur because several tests try to translate the same content at the same time.
As a result, one of these tests succeeds and the next one fails because the state is already "machine translated".

Before the introduced changes, these tests seemed more stable because simultaneous translation of the same entity was possible.

**Comment by timoludwig** [comment]
> As far as I understand, these failures occur because several tests try to translate the same content at the same time. As a result, one of these tests succeeds and the next one fails because the state is already "machine translated".
> 
> Before the introduced changes, these tests seemed more stable because simultaneous translation of the same entity was possible, but now it's not.

Ah okay, thanks for the explanation. But in this case we could also define a mapping of page ids per role, such as they don't interfere with each other? And also use a completely different page (which is not translated during the `test_auto_translate_easy_german` test) for the error-check in `test_summ_ai_error_handling`?
Also, I think it would be cool to test the feature of this PR itself, meaning that we validate that really an error occurs when an already translated page is translated again.

**Review by seluianova — COMMENTED** [review]
Fixed failing tests + added additional tests to check errors:
- There already is an up-to-date translation for {}
- All the selected translations are already up-to-date.

**Review by seluianova — APPROVED** [review]
My change request has been addressed, so I am adding an approval from my side

**Review by MizukiTemma — APPROVED** [review]
Looks good 😃 Thank you 🎉

**Pull request merged**
